## German used cars data
In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

# *dateCrawled* - When this ad was first crawled. All field-values are taken from this date.
# *name* - Name of the car.
# *seller* - Whether the seller is private or a dealer.
# *offerType* - The type of listing
# *price* - The price on the ad to sell the car.
# *abtest* - Whether the listing is included in an A/B test.
# *vehicleType* - The vehicle Type.
# *yearOfRegistration* - The year in which which year the car was first registered.
# *gearbox* - The transmission type.
# *powerPS* - The power of the car in PS.
# *model* - The car model name.
# *kilometer* - How many kilometers the car has driven.
# *monthOfRegistration* - The month in which which year the car was first registered.
# *fuelType* - What type of fuel the car uses.
# *brand* - The brand of the car.
# *notRepairedDamage* - If the car has a damage which is not yet repaired.
# *dateCreated* - The date on which the eBay listing was created.
# *nrOfPictures* - The number of pictures in the ad.
# *postalCode* - The postal code for the location of the vehicle.
# *lastSeenOnline* - When the crawler saw this ad last online.

In [5]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [6]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [7]:
autos.info()
autos.head(20)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15  2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16  2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18  2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19  2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   

    seller offerType    price   abtest vehicleType  yearOfRegistration  \
0   privat   Angebot   $5,000  control         bus                2004   
1   privat   Angebot   $8,500  control   limousine                1997   
2   privat   Angebot   $8,990     test   limousine                2009   
3   privat   Angebot   $4,350  control  kleinwagen                2007   
4   privat   Angebot   $1,350     test       kombi                2003   
5   privat   Angebot   $7,900     test         bus                2006   
6   privat   Angebot     $300     test   limousine                1995   
7   privat   Angebot   $1,990  control   limousine                1998   
8   privat   Angebot     $250     test         NaN                2000   
9   privat   Angebot     $590  control         bus                1997   
10  privat   Angebot     $999     test         NaN                2017   
11  privat   Angebot     $350  control         NaN                2000   
12  privat   Angebot   $5,299  control  kleinwagen                2010   
13  privat   Angebot   $1,350  control   limousine                1999   
14  privat   Angebot   $3,999     test  kleinwagen                2007   
15  privat   Angebot  $18,900     test       coupe                1982   
16  privat   Angebot     $350     test       kombi                1999   
17  privat   Angebot   $5,500     test       coupe                1990   
18  privat   Angebot     $300  control         bus                1995   
19  privat   Angebot   $4,150  control         suv                2004   

      gearbox  powerPS     model   odometer  monthOfRegistration fuelType  \
0     manuell      158    andere  150,000km                    3      lpg   
1   automatik      286       7er  150,000km                    6   benzin   
2     manuell      102      golf   70,000km                    7   benzin   
3   automatik       71    fortwo   70,000km                    6   benzin   
4     manuell        0     focus  150,000km                    7   benzin   
5   automatik      150   voyager  150,000km                    4   diesel   
6     manuell       90      golf  150,000km                    8   benzin   
7     manuell       90      golf  150,000km                   12   diesel   
8     manuell        0     arosa  150,000km                   10      NaN   
9     manuell       90    megane  150,000km     

# There are some columns with non-null values. Not repaired damage has the most.
## some integer types. Some can be converted to integer/float from object
## German words used

In [8]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [9]:
autos = autos.rename(columns={"yearOfRegistration": "registration_year", "monthOfRegistration": "registration_month", "notRepairedDamage": "unrepaired_damage", "dateCreated": "ad_created", "offerType": "offer_type", "fuelType": "fuel_type", "dateCrawled": "date_crawled", "nrOfPictures":"nr_of_pictures", "postalCode":"postal_code", "lastSeen":"last_seen", "vehicleType": "vehicle_type", "powerPS" : "power_ps" })
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## camelCase to snake_case

In [10]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

# observations
## price and odometer - string to numeric
## two columns have 49999 counts of same data, some have more than 30000 counts, namely - seller,offer_type- 49999 + 30000+ - gearbox,odometer,fuel_type, unrepaired damage
## some have only two unique values

In [11]:
odometer = autos["odometer"].str.replace("km","")
odometer = odometer.str.replace(",","")
autos["odometer"] = odometer.astype(int)
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)
price = autos["price"].str.replace("$","")
price = price.str.replace(",","").astype(int)
autos["price"] = price


In [12]:
autos = autos.drop(["seller","offer_type","nr_of_pictures"], axis=1)

In [13]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [15]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
print(autos["price"].value_counts().sort_index(ascending= False).head(20))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


In [16]:
print(autos["price"].value_counts().sort_index(ascending= True).head(20))

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


## Observations
1421 values = 0. We will remove them. While other low values are actually possible as Ebay is an auction site. In high-end, values steadily increase till 35000, then shoot up.with only some outliers.

In [17]:
autos["price"] = autos.loc[(autos["price"]>0) & (autos["price"]<35001), "price"]

In [18]:
autos["price"].value_counts().sort_index().head()

1.0    156
2.0      3
3.0      1
5.0      2
8.0      1
Name: price, dtype: int64

In [19]:
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [20]:
print(autos["odometer_km"].value_counts())

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


# odometer_km doesn't seem to have outliers.values are rounded.

# DateColumns
## date_crawled, ad_created, last_seen are in timestamp format, first10 letters are dates.

In [21]:
print(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))
print(autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))
print(autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True))


2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64
2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
2015-12-30    0.00002
2016-01-03    0.00002
2016-01-07    0.00002
2016-01-10    0.00004
2016-01-13    0.000

In [22]:
print(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_values())


2016-04-07    0.00142
2016-04-06    0.00318
2016-03-18    0.01306
2016-04-05    0.01310
2016-03-06    0.01394
2016-03-13    0.01556
2016-03-05    0.02538
2016-03-24    0.02910
2016-03-16    0.02950
2016-03-27    0.03104
2016-03-17    0.03152
2016-03-25    0.03174
2016-03-31    0.03192
2016-03-10    0.03212
2016-03-23    0.03238
2016-03-11    0.03248
2016-03-26    0.03248
2016-03-22    0.03294
2016-03-09    0.03322
2016-03-08    0.03330
2016-03-30    0.03362
2016-04-01    0.03380
2016-03-15    0.03398
2016-03-29    0.03418
2016-03-28    0.03484
2016-03-19    0.03490
2016-04-02    0.03540
2016-03-07    0.03596
2016-04-04    0.03652
2016-03-14    0.03662
2016-03-12    0.03678
2016-03-21    0.03752
2016-03-20    0.03782
2016-04-03    0.03868
Name: date_crawled, dtype: float64


## The site was crawled between 5th March and 7th April 2016. The percentages are similar for most dates.

In [23]:
print(autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_values())


2015-12-30    0.00002
2015-08-10    0.00002
2015-11-10    0.00002
2016-02-07    0.00002
2016-01-03    0.00002
2016-02-16    0.00002
2015-09-09    0.00002
2016-01-29    0.00002
2016-02-22    0.00002
2016-01-07    0.00002
2016-02-11    0.00002
2016-01-13    0.00002
2016-02-17    0.00002
2016-02-08    0.00002
2015-12-05    0.00002
2016-01-22    0.00002
2015-06-11    0.00002
2016-01-14    0.00002
2016-01-16    0.00002
2016-02-01    0.00002
2016-02-24    0.00004
2016-02-18    0.00004
2016-02-09    0.00004
2016-01-10    0.00004
2016-02-20    0.00004
2016-02-14    0.00004
2016-02-02    0.00004
2016-02-05    0.00004
2016-02-26    0.00004
2016-02-25    0.00006
               ...   
2016-03-06    0.01512
2016-03-13    0.01692
2016-03-05    0.02304
2016-03-24    0.02908
2016-03-16    0.03000
2016-03-27    0.03090
2016-03-17    0.03120
2016-03-10    0.03186
2016-03-25    0.03188
2016-03-31    0.03192
2016-03-23    0.03218
2016-03-26    0.03256
2016-03-11    0.03278
2016-03-22    0.03280
2016-03-09

In [24]:
print(autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index().tail(50))


2016-02-18    0.00004
2016-02-19    0.00006
2016-02-20    0.00004
2016-02-21    0.00006
2016-02-22    0.00002
2016-02-23    0.00008
2016-02-24    0.00004
2016-02-25    0.00006
2016-02-26    0.00004
2016-02-27    0.00012
2016-02-28    0.00020
2016-02-29    0.00016
2016-03-01    0.00010
2016-03-02    0.00010
2016-03-03    0.00086
2016-03-04    0.00144
2016-03-05    0.02304
2016-03-06    0.01512
2016-03-07    0.03474
2016-03-08    0.03334
2016-03-09    0.03324
2016-03-10    0.03186
2016-03-11    0.03278
2016-03-12    0.03662
2016-03-13    0.01692
2016-03-14    0.03522
2016-03-15    0.03374
2016-03-16    0.03000
2016-03-17    0.03120
2016-03-18    0.01372
2016-03-19    0.03384
2016-03-20    0.03786
2016-03-21    0.03772
2016-03-22    0.03280
2016-03-23    0.03218
2016-03-24    0.02908
2016-03-25    0.03188
2016-03-26    0.03256
2016-03-27    0.03090
2016-03-28    0.03496
2016-03-29    0.03414
2016-03-30    0.03344
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03508
2016-04-03

## Ads were created most times between march 5th to April 5th with different percentages but similar range. On dates other than these, percentages seem to be very less.

In [25]:
print(autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_values())


2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-18    0.00742
2016-03-08    0.00760
2016-03-13    0.00898
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-14    0.01280
2016-03-19    0.01574
2016-03-15    0.01588
2016-03-27    0.01602
2016-03-16    0.01644
2016-03-26    0.01696
2016-03-23    0.01858
2016-03-25    0.01920
2016-03-24    0.01956
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-28    0.02086
2016-03-22    0.02158
2016-03-29    0.02234
2016-04-01    0.02310
2016-03-12    0.02382
2016-03-31    0.02384
2016-04-04    0.02462
2016-03-30    0.02484
2016-04-02    0.02490
2016-04-03    0.02536
2016-03-17    0.02792
2016-04-05    0.12428
2016-04-07    0.13092
2016-04-06    0.22100
Name: last_seen, dtype: float64


# the last 3 days seem to have abnormally high percentage of last seen values.all other values have a different range altogether. these could be erroneous. 

In [26]:
print(autos["registration_year"].describe())

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


## Max year is 9999,while min year is 1000.these are erroneous values.
Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic. We calculate percentages

In [27]:
autos["registration_year"].between(1900,2016).value_counts(normalize = True) 

True     0.96056
False    0.03944
Name: registration_year, dtype: float64

We use between() to find out the percentage of truth values for data being between 1900-2016. We can safely ignore other values as it is outside 1900-2016 for only about 4% ofthe values. 

In [28]:
autos["registration_year"].value_counts(normalize = True).sort_values().tail(25).sort_index() 

1993    0.00890
1994    0.01320
1995    0.02626
1996    0.02888
1997    0.04056
1998    0.04906
1999    0.06000
2000    0.06708
2001    0.05406
2002    0.05066
2003    0.05454
2004    0.05474
2005    0.06030
2006    0.05416
2007    0.04608
2008    0.04462
2009    0.04196
2010    0.03194
2011    0.03268
2012    0.02646
2013    0.01612
2014    0.01332
2016    0.02632
2017    0.02906
2018    0.00984
Name: registration_year, dtype: float64

looks like the last 22 years have seen most percentage of registrations 1994 - 2016 precisely. While there are many erroneous entries for 2017.

In [29]:
print(autos["brand"].unique())
print(autos["brand"].unique().shape)

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']
(40,)


In [30]:
print(autos["brand"].value_counts(normalize = True))

volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
smart             0.01402
citroen           0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
mitsubishi        0.00812
honda             0.00798
kia               0.00712
alfa_romeo        0.00658
porsche           0.00588
suzuki            0.00586
chevrolet         0.00566
chrysler          0.00362
dacia             0.00258
daihatsu          0.00256
jeep              0.00220
subaru            0.00218
land_rover        0.00198
saab              0.00160
daewoo            0.00158
trabant           0.00156
jaguar            0.00154
rover             0.00138
lancia      

From brand data in percentages itcanbefoundoutthat 17 brands are present more than 1%. So we will chose them foraggregation, keeping 1% as the criteria.

In [31]:
unique_brands = autos["brand"].value_counts().head(17).index

In [32]:
print(unique_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'sonstige_autos'],
      dtype='object')


In [33]:
mean_price_top_17_brands ={}
for s in unique_brands:
    values_price = autos.loc[autos["brand"] == s,"price"]
    mean_price = values_price.mean()
    mean_price_top_17_brands[s] = mean_price
print(mean_price_top_17_brands)

{'peugeot': 3065.611888111888, 'skoda': 6353.544871794872, 'opel': 2937.7755875663383, 'mazda': 4011.849593495935, 'nissan': 4669.3859649122805, 'audi': 8273.284132841329, 'ford': 3562.143534994069, 'smart': 3518.102305475504, 'mercedes_benz': 7620.358726673985, 'fiat': 2793.8700475435817, 'citroen': 3761.8950437317785, 'bmw': 7679.428626320845, 'renault': 2392.224612736661, 'volkswagen': 5215.609569099379, 'sonstige_autos': 8034.256818181818, 'seat': 4315.744565217391, 'toyota': 5075.295081967213}


# Now, only top 5.


In [34]:
top6_brands = autos["brand"].value_counts().head(6).index
print(top6_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [35]:
mean_price_top6_brands ={}
for s in top6_brands:
    values_price = autos.loc[autos["brand"] == s,"price"]
    mean_price = values_price.mean()
    mean_price_top6_brands[s] = int(mean_price)
print(mean_price_top6_brands)

{'opel': 2937, 'ford': 3562, 'audi': 8273, 'volkswagen': 5215, 'mercedes_benz': 7620, 'bmw': 7679}


# Top 5 brands and mean price
We aggregated across brands to understand mean price. We observed that in the top 5 brands, there's a distinct price gap.

Audi, BMW and Mercedes Benz are more expensive.
Opel and Ford are less expensive.
Volkswagen is in between.

# For the top 5 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

# - it's difficult to compare more than two aggregate series objects if we want to extend to more columns
# - we can't compare more than a few rows from each series object
# - we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To do this, we'll need to learn two pandas methods:

# - pandas series constructor
# - pandas dataframe constructor

In [36]:
brand_mean_mileage = {}
for s in top6_brands:
    values_mileage = autos.loc[autos["brand"] == s,"odometer_km"]
    mean_mileage = values_mileage.mean()
    brand_mean_mileage[s] = int(mean_mileage)
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(mean_price_top6_brands).sort_values(ascending=False)












In [39]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info["mean_price"]



In [40]:
brand_info

mean_mileage  mean_price
0           NaN         NaN
1           NaN         NaN